src/server/shell/basic_accessibility_manager.h

```diff
@@ -86,15 +88,63 @@ class BasicAccessibilityManager : public AccessibilityManager
         std::shared_ptr<MousekeyPointer> mousekey_pointer;
     };
 
+    template <typename Transformer> class Registration
+    {
+    public:
+        Registration(
+            std::shared_ptr<Transformer> const& transformer,
+            std::shared_ptr<input::InputEventTransformer> const& event_transformer) :
+            transformer{transformer},
+            event_transformer{event_transformer}
+        {
+        }
+
+        ~Registration()
+        {
+            remove_registration();
+        }
+
+        void add_registration()
+        {
+            assert(!registration.has_value());
```

Comment:
1. Do not put implementation in header files without a good reason: it makes for fragile ABIs and increases compile times.
3. Do not use `assert()` to test preconditions: if they are worth testing, then they should (mostly) be tested regardless of `NDEBUG`.